src/controller/DeviceUnpair.cpp

```diff
@@ -0,0 +1,187 @@
+        ReadCurrentFabricIndex(proxy);
+        break;
+    }
+    case kSendRemoveFabric: {
+        SendRemoveFabricIndex(proxy, mFabricIndex);
+        break;
+    }
+    case kUnpairError:
+    case kUnpairCleanUp:
+    default: {
+        FinishUnpairDevice(err);
+        break;
+    }
+    }
+}
+
+void DeviceUnpair::ReadCurrentFabricIndex(DeviceProxy * proxy)
+{
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
+    app::ReadPrepareParams readParams(proxy->GetSecureSession().Value());
+
+    app::AttributePathParams readPaths[1];
+    readPaths[0] = app::AttributePathParams(OperationalCredentials::Id, OperationalCredentials::Attributes::CurrentFabricIndex::Id);
+
+    readParams.mpAttributePathParamsList    = readPaths;
+    readParams.mAttributePathParamsListSize = 1;
+    readParams.mIsFabricFiltered            = false;
+
+    auto attributeCache = Platform::MakeUnique<app::ClusterStateCache>(*this);
+    auto readClient     = Platform::MakeUnique<app::ReadClient>(
+        engine, proxy->GetExchangeManager(), attributeCache->GetBufferedCallback(), app::ReadClient::InteractionType::Read);
+    CHIP_ERROR err = readClient->SendRequest(readParams);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Controller, "Failed to send read request for networking clusters");
+        PerformDeviceUnpairStep(proxy, UnpairDeviceStage::kUnpairError, err);
+        return;
+    }
+
+    mAttributeCache = std::move(attributeCache);
+    mReadClient     = std::move(readClient);
+}
+
+void DeviceUnpair::SendRemoveFabricIndex(DeviceProxy * proxy, FabricIndex fabricIndex)
+{
+    if (fabricIndex == kUndefinedFabricIndex)
+    {
+        PerformDeviceUnpairStep(nullptr, UnpairDeviceStage::kUnpairError, CHIP_ERROR_INVALID_FABRIC_INDEX);
+        return;
+    }
+    ChipLogProgress(Controller, "SendRemoveFabricIndex : %u", fabricIndex);
+
+    OperationalCredentials::Commands::RemoveFabric::Type request;
+    request.fabricIndex = fabricIndex;
+
+    OperationalCredentialsCluster cluster(*proxy->GetExchangeManager(), proxy->GetSecureSession().Value(), 0);
+
+    cluster.InvokeCommand(request, this, OnRemoveFabric, OnCommandFailure);
+}
+
+void DeviceUnpair::FinishUnpairDevice(CHIP_ERROR err)
+{
+    mAttributeCache         = nullptr;
+    mReadClient             = nullptr;
+    mProxy                  = nullptr;
+    mOperationalDeviceProxy = nullptr;
+
+    if (mCallback != nullptr)
+    {
+        mCallback->OnDeviceUnpair(mRemoteDeviceId, err);
+    }
+}
+
+void DeviceUnpair::OnDeviceConnectedFn(void * context, Messaging::ExchangeManager & exchangeMgr, SessionHandle & sessionHandle)
+{
+    DeviceUnpair * deviceUnpair = static_cast<DeviceUnpair *>(context);
+    VerifyOrReturn(deviceUnpair != nullptr, ChipLogProgress(Controller, "Device connected callback with null context. Ignoring"));
+
+    auto proxy                            = Platform::MakeShared<OperationalDeviceProxy>(&exchangeMgr, sessionHandle);
```

Comment:
This does raise the question of why we take a DeviceProxy to start with....

Reply:
When operating the chiptool, I created a session and created a Device Proxy because I need to re-engage the CASE.

Reply:
I don't understand this part.  You create a session, fine.  You can pass the session to this process here.  It does not need a DeviceProxy, which is deprecated and should not be added in new code.